Error while checking build/O2/o2-dev-fairroot for 4c8fd37ff26c35aced6695fa57286179ed4c1018:
```
sw/BUILD/FairRoot-latest/log
-- Checking C++11 support for "decltype": works
-- Checking C++11 support for "initializer_list"
-- Checking C++11 support for "initializer_list": works
-- Checking C++11 support for "lambda"
-- Checking C++11 support for "lambda": works
-- Checking C++11 support for "long_long"
-- Checking C++11 support for "long_long": works
-- Checking C++11 support for "nullptr"
-- Checking C++11 support for "nullptr": works
-- Checking C++11 support for "regex"
-- Checking C++11 support for "regex": works
-- Checking C++11 support for "rvalue-references"
-- Checking C++11 support for "rvalue-references": works
-- Checking C++11 support for "sizeof_member"
-- Checking C++11 support for "sizeof_member": works
-- Checking C++11 support for "static_assert"
-- Checking C++11 support for "static_assert": works
-- Checking C++11 support for "variadic_templates"
-- Checking C++11 support for "variadic_templates": works
-- Checking C++11 support for "sharedpointer"
-- Checking C++11 support for "sharedpointer": works
-- Checking C++11 support for "funcptr-to-lambda-conversion"
-- Checking C++11 support for "funcptr-to-lambda-conversion": works
-- fairsoft-config not found
--- Found a Linux system
--- Found GNU compiler collection
--- Build Type: RELWITHDEBINFO
--- Compiler Flags: -fPIC -O2 -std=c++14  -O2 -g -Wshadow
-- Looking for Root...
-- Looking for Root... - Found /mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/ROOT/v6-12-06-2/bin/root
-- Looking for Root... - Found version is 6.12/06 
CMake Error at CMakeLists.txt:162 (find_package):
  By not providing "FindFairLogger.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "FairLogger", but CMake did not find one.

  Could not find a package configuration file provided by "FairLogger"
  (requested version 1.1.0) with any of the following names:

    FairLoggerConfig.cmake
    fairlogger-config.cmake

  Add the installation prefix of "FairLogger" to CMAKE_PREFIX_PATH or set
  "FairLogger_DIR" to a directory containing one of the above files.  If
  "FairLogger" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
```
Full log [here](https://ali-ci.cern.ch/repo/logs/AliceO2Group/AliceO2/1074/4c8fd37ff26c35aced6695fa57286179ed4c1018/build_O2_o2-dev-fairroot/fullLog.txt).

--- a/Steer/DigitizerWorkflow/CMakeLists.txt
+++ b/Steer/DigitizerWorkflow/CMakeLists.txt
@@ -25,6 +25,7 @@ O2_GENERATE_EXECUTABLE(
   src/SimReaderSpec.cxx
   src/CollisionTimePrinter.cxx
   src/TPCDriftTimeDigitizerSpec.cxx
+  src/TPCDigitRootWriterSpec.cxx
 
   BUCKET_NAME ${MODULE_BUCKET_NAME}
 )

--- a/Steer/DigitizerWorkflow/src/SimReaderSpec.cxx
+++ b/Steer/DigitizerWorkflow/src/SimReaderSpec.cxx
@@ -42,6 +42,25 @@ DataProcessorSpec getSimReaderSpec(int fanoutsize, std::shared_ptr<std::vector<i
 
     // counter to make sure we are sending the data only once
     static int counter = 0;
+
+    static bool finished = false;
+    if (finished) {
+      // we need to send this in a different time slice
+      // send message telling tpc workers that they can terminate
+      for (const auto& channel : *tpcsubchannels.get()) {
+        // -1 is marker for end of work
+        pc.outputs().snapshot(
+          Output{ "SIM", "TPCSECTORASSIGN", static_cast<SubSpecificationType>(channel), Lifetime::Condition }, -1);
+        // not sure if I have to resend this as well? Seems not necessary
+        pc.outputs().snapshot(
+          Output{ "SIM", "COLLISIONCONTEXT", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+          context);
+      }
+      // do this only one
+      pc.services().get<ControlService>().readyToQuit(false);
+      return;
+    }
+
     if (counter++ == 0) {
       LOG(INFO) << "SENDING " << eventrecords.size() << " records";
 
@@ -59,16 +78,6 @@ DataProcessorSpec getSimReaderSpec(int fanoutsize, std::shared_ptr<std::vector<i
           context);
         tpcchannelcounter++;
       }
-      // send message telling tpc workers that they can terminate
-      for (const auto& channel : *tpcsubchannels.get()) {
-        // -1 is marker for end of work
-        pc.outputs().snapshot(
-          Output{ "SIM", "TPCSECTORASSIGN", static_cast<SubSpecificationType>(channel), Lifetime::Condition }, -1);
-        // not sure if I have to resend this as well? Seems not necessary
-        pc.outputs().snapshot(
-          Output{ "SIM", "COLLISIONCONTEXT", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
-          context);
-      }
 
       // everything not done previously treat here
       for (int subchannel = 0; subchannel < fanoutsize; ++subchannel) {
@@ -80,9 +89,7 @@ DataProcessorSpec getSimReaderSpec(int fanoutsize, std::shared_ptr<std::vector<i
           Output{ "SIM", "COLLISIONCONTEXT", static_cast<SubSpecificationType>(subchannel), Lifetime::Timeframe },
           context);
       }
-
-      // do this only one
-      pc.services().get<ControlService>().readyToQuit(false);
+      finished = true;
     }
   };
 

--- a/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx
+++ b/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx
@@ -16,6 +16,7 @@
 
 // for TPC
 #include "TPCDriftTimeDigitizerSpec.h"
+#include "TPCDigitRootWriterSpec.h"
 #include "TPCBase/Sector.h"
 
 #include <cstdlib>
@@ -98,5 +99,7 @@ void defineDataProcessing(WorkflowSpec& specs)
     fanoutsize++;
   }
 
+  // for writing digits to disc
+  specs.emplace_back(o2::TPC::getTPCDigitRootWriterSpec(lanes));
   specs.emplace_back(o2::steer::getSimReaderSpec(fanoutsize, tpcsectors, tpclanes));
 }

--- a/Steer/DigitizerWorkflow/src/TPCDigitRootWriterSpec.cxx
+++ b/Steer/DigitizerWorkflow/src/TPCDigitRootWriterSpec.cxx
@@ -1,0 +1,209 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   TPCDigitRootFileWriterSpec.cxx
+/// @author Matthias Richter, Sandro Wenzel
+/// @since  2018-04-19
+/// @brief  Processor spec for a ROOT file writer for TPC digits
+
+#include "TPCDigitRootWriterSpec.h"
+#include "Framework/CallbackService.h"
+#include "Framework/ControlService.h"
+#include "TPCBase/Sector.h"
+#include "TPCBase/Digit.h"
+#include <SimulationDataFormat/MCCompLabel.h>
+#include <SimulationDataFormat/MCTruthContainer.h>
+#include <TFile.h>
+#include <TTree.h>
+#include <TBranch.h>
+#include <memory> // for make_shared, make_unique, unique_ptr
+#include <stdexcept>
+#include <sstream>
+#include <string>
+#include <vector>
+
+using namespace o2::framework;
+using SubSpecificationType = o2::framework::DataAllocator::SubSpecificationType;
+
+namespace o2
+{
+namespace TPC
+{
+
+template <typename T>
+TBranch* getOrMakeBranch(TTree& tree, const char* basename, int sector, T* ptr)
+{
+  std::stringstream stream;
+  stream << basename << "_" << sector;
+  const auto brname = stream.str().c_str();
+  if (auto br = tree.GetBranch(brname)) {
+    br->SetAddress(static_cast<void*>(&ptr));
+    return br;
+  }
+  // otherwise make it
+  return tree.Branch(brname, ptr);
+}
+
+/// create the processor spec
+/// describing a processor aggregating digits for various TPC sectors and writing them to file
+/// MC truth information is also aggregated and written out
+DataProcessorSpec getTPCDigitRootWriterSpec(int numberofsourcedevices)
+{
+  // assign input names to each channel
+  auto digitchannelname = std::make_shared<std::vector<std::string>>();
+  auto labelchannelname = std::make_shared<std::vector<std::string>>();
+  auto sectorchannelname = std::make_shared<std::vector<std::string>>();
+  for (int i = 0; i < numberofsourcedevices; ++i) {
+    {
+      std::stringstream ss;
+      ss << "digitinput" << i;
+      digitchannelname->push_back(ss.str());
+    }
+    {
+      std::stringstream ss;
+      ss << "sectorinput" << i;
+      sectorchannelname->push_back(ss.str());
+    }
+    {
+      std::stringstream ss;
+      ss << "labelinput" << i;
+      labelchannelname->push_back(ss.str());
+    }
+  }
+
+  auto initFunction = [numberofsourcedevices, digitchannelname, sectorchannelname, labelchannelname](InitContext& ic) {
+    // get the option from the init context
+    auto filename = ic.options().get<std::string>("tpc-digit-outfile");
+    auto treename = ic.options().get<std::string>("treename");
+
+    auto outputfile = std::make_shared<TFile>(filename.c_str(), "RECREATE");
+    auto outputtree = std::make_shared<TTree>(treename.c_str(), treename.c_str());
+
+    // container for incoming digits + label
+    auto digits = std::make_shared<std::vector<o2::TPC::Digit>>();
+    auto labels = std::make_shared<o2::dataformats::MCTruthContainer<o2::MCCompLabel>>();
+
+    // the callback to be set as hook at stop of processing for the framework
+    auto finishWriting = [outputfile, outputtree]() {
+      // check/verify number of entries (it should be same in all branches)
+
+      // will return a TObjArray
+      const auto brlist = outputtree->GetListOfBranches();
+      int entries = -1; // init to -1 (as unitialized)
+      for (TObject* entry : *brlist) {
+        auto br = static_cast<TBranch*>(entry);
+        int brentries = br->GetEntries();
+        if (entries == -1) {
+          entries = brentries;
+        } else {
+          if (brentries != entries) {
+            LOG(WARNING) << "INCONSISTENT NUMBER OF ENTRIES IN BRANCHES " << entries << " vs " << brentries;
+            entries = brentries;
+          }
+        }
+      }
+      if (entries > 0) {
+        outputtree->SetEntries(entries);
+      }
+      outputtree->Write();
+      outputfile->Close();
+    };
+    ic.services().get<CallbackService>().set(CallbackService::Id::Stop, finishWriting);
+
+    // set up the processing function
+    // using by-copy capture of the worker instance shared pointer
+    // the shared pointer makes sure to clean up the instance when the processing
+    // function gets out of scope
+    auto processingFct = [outputfile, outputtree, digits, digitchannelname, sectorchannelname, labelchannelname,
+                          numberofsourcedevices](ProcessingContext& pc) {
+      static bool finished = false;
+      if (finished) {
+        // avoid being executed again when marked as finished;
+        return;
+      }
+
+      static int finishchecksum = 0;
+
+      auto isComplete = [numberofsourcedevices](int i) {
+        if (i == numberofsourcedevices * (numberofsourcedevices - 1) / 2) {
+          return true;
+        }
+        return false;
+      };
+
+      for (int d = 0; d < numberofsourcedevices; ++d) {
+        const auto cname = digitchannelname->operator[](d);
+        const auto sname = sectorchannelname->operator[](d);
+        const auto lname = labelchannelname->operator[](d);
+
+        auto sectorptr = pc.inputs().get<int>(sname.c_str());
+        const int sector = *sectorptr.get();
+        LOG(INFO) << "GOT DIGITS FOR SECTOR " << sector;
+
+        if (sector == -1) {
+          finishchecksum += d;
+          if (isComplete(finishchecksum)) {
+            finished = true;
+            pc.services().get<ControlService>().readyToQuit(false);
+            return;
+          }
+        } else {
+          // have to do work ...
+          // the digits
+          auto indata = pc.inputs().get<std::vector<o2::TPC::Digit>>(cname.c_str());
+          LOG(INFO) << "DIGIT SIZE " << indata.size();
+          *digits.get() = std::move(indata);
+          {
+            // connect this to a particular branch
+            auto br = getOrMakeBranch(*outputtree.get(), "TPCDigit", sector, digits.get());
+            br->Fill();
+            br->ResetAddress();
+          }
+
+          // the labels
+          auto labeldata = pc.inputs().get<o2::dataformats::MCTruthContainer<o2::MCCompLabel>>(lname.c_str());
+          auto labeldataRaw = labeldata.get();
+          {
+            auto br = getOrMakeBranch(*outputtree.get(), "TPCDigitMCTruth", sector, &labeldataRaw);
+            br->Fill();
+            br->ResetAddress();
+          }
+        }
+      }
+    };
+
+    // return the actual processing function as a lambda function using variables
+    // of the init function
+    return processingFct;
+  };
+
+  std::vector<InputSpec> inputs;
+  for (int d = 0; d < numberofsourcedevices; ++d) {
+    inputs.emplace_back(InputSpec{ (*digitchannelname.get())[d].c_str(), "TPC", "DIGITS",
+                                   static_cast<SubSpecificationType>(d), Lifetime::Timeframe }); // digit input
+    inputs.emplace_back(InputSpec{ (*labelchannelname.get())[d].c_str(), "TPC", "DIGITSMCTR",
+                                   static_cast<SubSpecificationType>(d), Lifetime::Timeframe });
+    inputs.emplace_back(InputSpec{ (*sectorchannelname.get())[d].c_str(), "TPC", "SECTOR",
+                                   static_cast<SubSpecificationType>(d), Lifetime::Timeframe });
+  }
+
+  return DataProcessorSpec{
+    "TPCDigitWriter",
+    inputs,
+    {}, // no output
+    AlgorithmSpec(initFunction),
+    Options{
+      { "tpc-digit-outfile", VariantType::String, "tpcdigits.root", { "Name of the input file" } },
+      { "treename", VariantType::String, "o2sim", { "Name of tree for tracks" } },
+    }
+  };
+}
+} // end namespace TPC
+} // end namespace o2

--- a/Steer/DigitizerWorkflow/src/TPCDigitRootWriterSpec.h
+++ b/Steer/DigitizerWorkflow/src/TPCDigitRootWriterSpec.h
@@ -1,0 +1,25 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef STEER_DIGITIZERWORKFLOW_SRC_TPCDIGITROOTWRITERSPEC_H_
+#define STEER_DIGITIZERWORKFLOW_SRC_TPCDIGITROOTWRITERSPEC_H_
+
+#include "Framework/DataProcessorSpec.h"
+
+namespace o2
+{
+namespace TPC
+{
+// numberofsourcedevices is the number of devices we receive digits from
+o2::framework::DataProcessorSpec getTPCDigitRootWriterSpec(int numberofsourcedevices = 1);
+}
+}
+
+#endif /* STEER_DIGITIZERWORKFLOW_SRC_TPCDIGITROOTWRITERSPEC_H_ */

--- a/Steer/DigitizerWorkflow/src/TPCDriftTimeDigitizerSpec.cxx
+++ b/Steer/DigitizerWorkflow/src/TPCDriftTimeDigitizerSpec.cxx
@@ -61,45 +61,56 @@ DataProcessorSpec getTPCDriftTimeDigitizer(int channel, bool cachehits)
   digitizertask->Init2();
 
   auto digitArray = std::make_shared<std::vector<o2::TPC::Digit>>();
+
   auto mcTruthArray = std::make_shared<o2::dataformats::MCTruthContainer<o2::MCCompLabel>>();
   // the task takes the ownership of digit array + mc truth array
   // TODO: make this clear in the API
   digitizertask->setOutputData(digitArray.get(), mcTruthArray.get());
 
-  auto doit = [simChain, simChains, digitizertask, digitArray, mcTruthArray](ProcessingContext& pc) {
+  auto doit = [simChain, simChains, digitizertask, digitArray, mcTruthArray, channel](ProcessingContext& pc) {
     static int callcounter = 0;
     callcounter++;
-
-    auto sectorptr = pc.inputs().get<int>("sectorassign");
-    if (sectorptr) {
-      LOG(INFO) << "GOT ASSIGNED SECTOR" << *sectorptr.get();
-    }
-    int sector = *sectorptr;
-
-    // no more tasks can be marked with a negative sector
-    if (sector == -1) {
-      // we are ready to quit
-      pc.services().get<ControlService>().readyToQuit(false);
+    static bool finished = false;
+    if (finished) {
       return;
     }
+
+    // extract which sector to treat (strangely this is a unique pointer)
+    auto sectorptr = pc.inputs().get<int>("sectorassign");
+    if (sectorptr) {
+      LOG(INFO) << "GOT ASSIGNED SECTOR " << *sectorptr.get();
+    }
+    int sector = *sectorptr;
 
     // ===| open file and register branches |=====================================
     // this is done at the moment for each worker function invocation
     // TODO: make this nicer or let the write service be handled outside
-    auto file = std::make_unique<TFile>(Form("tpc_digi_%i_instance%i.root", sector, callcounter), "recreate");
-    auto outtree = std::make_unique<TTree>("o2sim", "TPC digits");
-    outtree->SetDirectory(file.get());
     auto digitArrayRaw = digitArray.get();
     auto mcTruthArrayRaw = mcTruthArray.get();
-    auto digitBranch = outtree->Branch(Form("TPCDigit_%i", sector), &digitArrayRaw);
-    auto mcTruthBranch = outtree->Branch(Form("TPCDigitMCTruth_%i", sector), &mcTruthArrayRaw);
+
+    // no more tasks can be marked with a negative sector
+    if (sector == -1) {
+      // we are ready to quit
+
+      // notify channels further down the road that we are done (why do I have to send digitArray here????)
+      digitArrayRaw->clear();
+      mcTruthArrayRaw->clear();
+      pc.outputs().snapshot(Output{ "TPC", "DIGITS", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+                            *digitArrayRaw);
+      pc.outputs().snapshot(
+        Output{ "TPC", "DIGITSMCTR", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+        *mcTruthArrayRaw);
+      pc.outputs().snapshot(Output{ "TPC", "SECTOR", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+                            sector);
+      pc.services().get<ControlService>().readyToQuit(false);
+      finished = true;
+      return;
+    }
 
     // obtain collision contexts
     auto context = pc.inputs().get<o2::steer::RunContext>("collisioncontext");
     auto& timesview = context->getEventRecords();
     LOG(DEBUG) << "GOT " << timesview.size() << " COLLISSION TIMES";
-
-    // extract which sector to treat (strangely this is a unique pointer)
 
     // if there is nothing ... return
     if (timesview.size() == 0) {
@@ -137,6 +148,10 @@ DataProcessorSpec getTPCDriftTimeDigitizer(int channel, bool cachehits)
     hitvectorsleft.resize(maxnumberofentries, nullptr);
     hitvectorsright.resize(maxnumberofentries, nullptr);
 
+    // this is to accum digits from different drift times
+    std::vector<o2::TPC::Digit> digitAccum;
+    o2::dataformats::MCTruthContainer<o2::MCCompLabel> labelAccum;
+
     for (int drift = 1; drift <= ndrifts; ++drift) {
       auto starttime = (drift - 1) * TPCDRIFT;
       auto endtime = drift * TPCDRIFT;
@@ -163,12 +178,22 @@ DataProcessorSpec getTPCDriftTimeDigitizer(int channel, bool cachehits)
         digitizertask->setupSector(sector);
         digitizertask->Exec2("");
 
-        // write digits + MC truth
-        outtree->Fill();
+        std::copy(digitArrayRaw->begin(), digitArrayRaw->end(), std::back_inserter(digitAccum));
+        labelAccum.mergeAtBack(*mcTruthArrayRaw);
+
+        // NOTE: we would like to send it here in order to avoid copying/accumulating !!
       }
     }
-    outtree->SetDirectory(file.get());
-    file->Write();
+    // write digits + MC truth
+    pc.outputs().snapshot(Output{ "TPC", "DIGITS", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+                          digitAccum);
+    pc.outputs().snapshot(
+      Output{ "TPC", "DIGITSMCTR", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe }, labelAccum);
+    pc.outputs().snapshot(Output{ "TPC", "SECTOR", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+                          sector);
+
+    //# outtree->SetDirectory(file.get());
+    //# file->Write();
     timer.Stop();
     LOG(INFO) << "Digitization took " << timer.CpuTime() << "s";
   };
@@ -199,9 +224,10 @@ DataProcessorSpec getTPCDriftTimeDigitizer(int channel, bool cachehits)
                                          static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
                               InputSpec{ "sectorassign", "SIM", "TPCSECTORASSIGN",
                                          static_cast<SubSpecificationType>(channel), Lifetime::Condition } },
-    Outputs{
-      // define channel by triple of (origin, type id of data to be sent on this channel, subspecification)
-    },
+    Outputs{ // define channel by triple of (origin, type id of data to be sent on this channel, subspecification)
+             OutputSpec{ "TPC", "DIGITS", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+             OutputSpec{ "TPC", "DIGITSMCTR", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe },
+             OutputSpec{ "TPC", "SECTOR", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe } },
     AlgorithmSpec{ initIt },
     Options{ { "simFile", VariantType::String, "o2sim.root", { "Sim (background) input filename" } },
              { "simFileS", VariantType::String, "", { "Sim (signal) input filename" } } }

